agent/lib/kontena/workers/weave_worker.rb

```diff
@@ -95,5 +69,51 @@ def weave_attach(container)
       error exc.backtrace.join("\n")
     end
 
+    # @param [String] container_id
+    # @param [String] overlay_cidr
+    def attach_overlay(container_id, overlay_cidr)
+      Actor[:network_adapter].exec(['--local', 'attach', overlay_cidr, '--rewrite-hosts', container.id])
+    end
+
+    # @param [Docker::Container]
+    def register_container_dns(container)
+      grid_name = container.labels['io.kontena.grid.name']
+      service_name = container.labels['io.kontena.service.name']
+      instance_number = container.labels['io.kontena.service.instance_number']
+      domain_name = container.config['Domainname'] || "#{grid_name}.kontena.local"
+      if container.default_stack?
+        hostname = container.labels['io.kontena.container.name']
+        dns_names = default_stack_dns_names(hostname, service_name, domain_name)
+        dns_names = dns_names + stack_dns_names(hostname, service_name, domain_name)
+
+      else
+        hostname = container.config['Hostname']
+        dns_names = stack_dns_names(hostname, service_name, domain_name)
+      end
+      dns_names.each do |name|
```

Comment:
Yes. The naming policy could be expressed nicely using something along the lines of `yield #{self.hostname}.#{self.domain}` and it would be easier to test.

Nothing we can't refactor later if we want to.